hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/audit/S3AAuditLogMergerAndParser.java

```diff
@@ -0,0 +1,546 @@
+
+    Path destFile = new Path(s3DestPath, "AuditLogFile");
+
+    List<HashMap<String, String>> entireAuditLogList = new ArrayList<>();
+    List<HashMap<String, String>> referrerHeaderList = new ArrayList<>();
+    List<HashMap<String, String>> auditLogList = new ArrayList<>();
+
+    try (FSDataOutputStream fsDataOutputStream = s3AFileSystem.create(
+        destFile)) {
+      // Iterating over the list of files to merge and parse
+      while (listOfS3LogFiles.hasNext()) {
+        Path logFile = listOfS3LogFiles.next().getPath();
+        FileStatus fileStatus = s3AFileSystem.getFileStatus(logFile);
+        int fileLength = (int) fileStatus.getLen();
+        byte[] byteBuffer = new byte[fileLength];
+        try (
+            FSDataInputStream fsDataInputStream = s3AFileSystem.open(logFile)) {
+
+          // Reads the file data in byte array
+          fsDataInputStream.readFully(byteBuffer);
+
+          // Convert byte array to string to make parsing easy
+          String byteStr = new String(byteBuffer, "UTF-8");
+          List<String> logs = IOUtils.readLines(new StringReader(byteStr));
+
+          // Parse each and every audit log from list of logs
+          for (String singleAuditLog : logs) {
+            // Parse audit log except referrer header
+            Map<String, String> auditLogMap = parseAuditLog(singleAuditLog);
+
+            String referrerHeader = auditLogMap.get("referrer");
+            if (referrerHeader == null || referrerHeader.equals("-")) {
+              LOG.info("Log didn't parsed : " + referrerHeader);
+              continue;
+            }
+
+            // Parse only referrer header
+            Map<String, String> referrerHeaderMap =
+                parseReferrerHeader(referrerHeader);
+            Map<String, String> entireAuditLogMap = new HashMap<>();
+            entireAuditLogMap.putAll(auditLogMap);
+            entireAuditLogMap.putAll(referrerHeaderMap);
+
+            // Adds every single map containing key-value pairs of single
+            // audit log into a list except referrer header key-value pairs.
+            auditLogList.add((HashMap<String, String>) auditLogMap);
+
+            // Also adds every single map containing key-value pairs of
+            // referrer header into a list.
+            referrerHeaderList.add((HashMap<String, String>) referrerHeaderMap);
+
+            // And adds every single map containing key-value pairs of entire
+            // audit log into a list including referrer header key-value pairs.
+            entireAuditLogList.add((HashMap<String, String>) entireAuditLogMap);
+          }
+        }
+        // Write byte array into a file in destination path.
+        fsDataOutputStream.write(byteBuffer);
+      }
+      // This method is used to convert the list of maps into avro file
+      // for querying using hive and spark.
+      convertToAvroFile(referrerHeaderList, auditLogList, s3AFileSystem,
```

Comment:
the avro record should be written incrementally too

Reply:
yes modified